File: MergeSort.py
def merge(arr,start,end,mid):
    size1 = mid-start+1
    size2 = end-mid
    
    arr1= [None]*size1
    arr2= [None]*size2
    
    for nums in range(size1):
        arr1[nums] = arr[start+nums]
    for nums in range(size2):
        arr2[nums] = arr[mid+nums+1]    
    
    print(arr1,arr2,"Inside")
    i,j,k=0,0,start
    while i<size1 and j<size2:
        
        if arr1[i]<arr2[j]:
            arr[k]=arr1[i]
            i+=1
            k+=1
        else:
            arr[k]=arr2[j]
            j+=1
            k+=1
    while i<size1:
        arr[k] = arr1[i]
        i+=1
        k+=1
    while j<size1:
        arr[k] = arr2[j]
        j+=1
        k+=1
    


def mergeDivison(arr,start,end):
    
    if start <end:
        mid = (start+end)//2
        mergeDivison(arr,start,mid)
        mergeDivison(arr,mid+1,end)
        
        print(arr,start,end,mid,"start end mid")
        merge(arr,start,end,mid)
        
        
    
list1 = [6,3,2,1]
mergeDivison(list1,0,len(list1)-1)
print(list1)

File: MergeSort.cpp
#include <iostream>

using namespace std;

void mergeSort(int arr[], int start, int mid, int end)
{
    int size1 = mid - start + 1;
    int size2 = end - mid;

    int temp1[size1];
    int temp2[size2];

    for (int i = 0; i < size1; i++)
    {
        temp1[i] = arr[start + i];
    }
    for (int i = 0; i < size2; i++)
    {
        temp2[i] = arr[mid + 1 + i];
    }
    int i = 0;
    int j = 0;
    int k = start;

    while (i < size1 && j < size2)
    {
        if (temp1[i] < temp2[j])
        {
            arr[k] = temp1[i];
            i++;
            k++;
        }
        else
        {
            arr[k] = temp2[j];
            j++;
            k++;
        }
    }
    while (i < size1)
    {
        arr[k] = temp1[i];
        i++;
        k++;
    }
    while (j < size2)
    {
        arr[k] = temp1[j];
        j++;
        k++;
    }
}

void mergeDivide(int arr[], int start, int end)
{
    if (start < end)
    {
        int mid = (start + end) / 2;
        mergeDivide(arr, start, mid);
        mergeDivide(arr, mid + 1, end);

        mergeSort(arr, start, mid, end);
    }
}
int main()
{
    int numbers[] = {8, 2, 5, 4};
    mergeDivide(numbers, 0, 3);
    // for (int i = 0; i < 4; i++)
    // {
    //     cout << numbers[i] << " ";
    // }
    // int r = 1;
    // int sum = r / 2;
    // cout << sum;
    return 0;
}